calyx/opt/src/passes/constant_port_prop.rs

```diff
@@ -0,0 +1,149 @@
+    /// name of comb group -> name of cell being invoked
+    comb_groups_to_modify: HashMap<ir::Id, ir::Id>,
+    /// comb groups used in while/if blocks, or used in multiple invokes
+    /// used to filter out comb groups that we *shouldn't* modify
+    /// NOTE: This needs to be in a separate set because we may process whiles/ifs before invokes.
+    comb_groups_used_elsewhere: HashSet<ir::Id>,
+}
+
+impl Named for ConstantPortProp {
+    fn name() -> &'static str {
+        "constant-port-prop"
+    }
+
+    fn description() -> &'static str {
+        "propagates constants when port values can be inferred"
+    }
+}
+
+impl Visitor for ConstantPortProp {
+    fn invoke(
+        &mut self,
+        s: &mut calyx_ir::Invoke,
+        _comp: &mut calyx_ir::Component,
+        _sigs: &LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        if let Some(cg) = &s.comb_group {
+            let cg_name = cg.borrow().name();
+            let cell_name = s.comp.borrow().name();
+            if let Some(registered_cell_name) =
+                self.comb_groups_to_modify.get(&cg_name)
+            {
+                if *registered_cell_name != cell_name {
+                    // there is a different invoke that is using the same comb group
+                    self.comb_groups_used_elsewhere.insert(cg_name);
+                }
+            } else {
+                // no invokes have used this comb group so far
+                self.comb_groups_to_modify.insert(cg_name, cell_name);
+            }
+        }
+        Ok(Action::Continue)
+    }
+
+    fn start_while(
+        &mut self,
+        s: &mut calyx_ir::While,
+        _comp: &mut calyx_ir::Component,
+        _sigs: &LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        if let Some(comb_group) = &s.cond {
+            self.comb_groups_used_elsewhere
+                .insert(comb_group.borrow().name());
+        }
+        Ok(Action::Continue)
+    }
+
+    fn start_if(
+        &mut self,
+        s: &mut calyx_ir::If,
+        _comp: &mut calyx_ir::Component,
+        _sigs: &LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        if let Some(comb_group) = &s.cond {
+            self.comb_groups_used_elsewhere
+                .insert(comb_group.borrow().name());
+        }
+        Ok(Action::Continue)
+    }
+
+    fn finish(
+        &mut self,
+        comp: &mut calyx_ir::Component,
+        sigs: &LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        let mut builder = ir::Builder::new(comp, sigs);
+        let one = builder.add_constant(1, 1);
+        let zero = builder.add_constant(0, 1);
+        // first, drop any comb groups that are used in while/ifs and in multiple invokes
+        for used_comb_group in &self.comb_groups_used_elsewhere {
+            self.comb_groups_to_modify.remove(used_comb_group);
+        }
```

Comment:
Definitely a nit, but you could use `comb_groups.iter().filter( ... )` instead
of directly mutating the hashmap if you prefer. I don't think it makes a big
difference either way though